File: desktop/src/services/vault/reviewScheme.test.ts
import { describe, it, expect } from 'vitest';
import {
  SCHEME_VAULT_PATH,
  applySchemeUpdate,
  buildSchemeFromPayload,
  cloneStore,
  normalizeQuestions,
  parseSchemeDoc,
  serializeSchemeDoc,
  DEFAULT_REVIEW_SCHEME_STORE,
  OFFICIAL_SCHEME_ID,
} from './reviewScheme';

describe('reviewScheme parse/serialize', () => {
  it('往返等价：parse(serialize(store)) 深相等', () => {
    const store = cloneStore(DEFAULT_REVIEW_SCHEME_STORE);
    const md = serializeSchemeDoc(store);
    const reparsed = parseSchemeDoc(md);
    expect(reparsed).toEqual(store);
  });

  it('序列化产出 ```json 块且含 activeSchemeId/schemes', () => {
    const md = serializeSchemeDoc(DEFAULT_REVIEW_SCHEME_STORE);
    expect(md).toContain('```json');
    expect(md).toContain('"activeSchemeId"');
    expect(md).toContain('"schemes"');
    expect(md).toContain(OFFICIAL_SCHEME_ID);
  });

  it('空/无 JSON 块/非法 JSON → 回退默认方案', () => {
    expect(parseSchemeDoc('')).toEqual(DEFAULT_REVIEW_SCHEME_STORE);
    expect(parseSchemeDoc('# 只有标题\n\n无 JSON')).toEqual(DEFAULT_REVIEW_SCHEME_STORE);
    expect(parseSchemeDoc('```json\n{不是合法json\n```')).toEqual(DEFAULT_REVIEW_SCHEME_STORE);
  });

  it('activeSchemeId 指向不存在的方案 → 回退首个方案', () => {
    const md = serializeSchemeDoc({
      activeSchemeId: '不存在',
      schemes: DEFAULT_REVIEW_SCHEME_STORE.schemes,
    });
    const parsed = parseSchemeDoc(md);
    expect(parsed.activeSchemeId).toBe(parsed.schemes[0].id);
  });

  it('保留 custom 方案稳定题目 ID（历史 rv: 可匹配）', () => {
    const store: typeof DEFAULT_REVIEW_SCHEME_STORE = {
      activeSchemeId: 'sch-x',
      schemes: [
        {
          id: 'sch-x',
          name: '我的方案',
          source: 'custom',
          isDefault: false,
          questions: [{ id: 'q-stable', title: '题', placeholder: 'p', required: true, maxLength: 500, sortOrder: 1 }],
        },
      ],
    };
    const reparsed = parseSchemeDoc(serializeSchemeDoc(store));
    expect(reparsed.schemes[0].questions[0].id).toBe('q-stable');
  });
});

describe('reviewScheme 规范化', () => {
  it('normalizeQuestions：补 sortOrder、截断 maxLength、过滤空标题、保留传入 id', () => {
    const questions = normalizeQuestions([
      { id: 'keep', title: '保留', placeholder: '', required: true, maxLength: 99999 },
      { title: '   ', placeholder: 'x', required: false }, // 空标题被过滤
    ]);
    expect(questions).toHaveLength(1);
    expect(questions[0].id).toBe('keep');
    expect(questions[0].sortOrder).toBe(1);
    expect(questions[0].maxLength).toBe(1000);
    expect(questions[0].placeholder).toBe('请在此输入你的思考...');
  });

  it('buildSchemeFromPayload：空名称/无题目返回 null，否则生成 custom 方案', () => {
    expect(buildSchemeFromPayload({ name: '', questions: [{ title: 'x', placeholder: '', required: true }] })).toBeNull();
    expect(
      buildSchemeFromPayload({ name: '方案', questions: [{ title: '', placeholder: '', required: true }] }),
    ).toBeNull();
    const scheme = buildSchemeFromPayload({
      name: '我的方案',
      questions: [{ title: '题一', placeholder: '', required: true }],
    });
    expect(scheme).not.toBeNull();
    expect(scheme!.source).toBe('custom');
    expect(scheme!.questions).toHaveLength(1);
  });

  it('applySchemeUpdate：official 方案不可改', () => {
    const official = DEFAULT_REVIEW_SCHEME_STORE.schemes.find((s) => s.source === 'official')!;
    expect(applySchemeUpdate(official, { id: official.id, name: '改名', questions: official.questions })).toBeNull();
  });

  it('SCHEME_VAULT_PATH 为 Reviews/scheme.md（vault 内 .md 才会被引擎同步）', () => {
    expect(SCHEME_VAULT_PATH).toBe('Reviews/scheme.md');
  });
});

File: mobile/test/features/review/review_scheme_test.dart
import 'package:flutter_test/flutter_test.dart';
import 'package:lifescale_mobile/features/review/data/review_scheme.dart';

/// 复盘方案解析测试：验证与桌面端 reviewScheme.ts 的契约一致性。
/// 关键不变量：official-* 题目 ID 永久稳定，保证历史 <!-- rv:questionId --> 可匹配。
void main() {
  group('parseSchemeDoc / serializeSchemeDoc', () {
    test('空文档 → 内置默认方案（official 4 题 + 轻量 3 题）', () {
      final store = parseSchemeDoc('');
      expect(store.activeSchemeId, officialSchemeId);
      expect(store.schemes, hasLength(2));
      final official = store.activeScheme;
      expect(official.source, 'official');
      expect(official.questions.map((q) => q.id), contains('official-done'));
      expect(official.questions.map((q) => q.id),
          contains('official-tomorrow'));
      expect(official.questions, hasLength(4));
    });

    test('异常 JSON → 回退默认方案，不抛异常', () {
      final store = parseSchemeDoc('```json\n{不是合法 json\n```');
      expect(store.activeSchemeId, officialSchemeId);
    });

    test('official 题目 ID 稳定（历史复盘可永久匹配）', () {
      final store = parseSchemeDoc('');
      final ids = store.activeScheme.questions.map((q) => q.id).toSet();
      expect(ids, containsAll(const [
        'official-done',
        'official-undone',
        'official-gain',
        'official-tomorrow',
      ]));
    });

    test('往返：serialize(parse(serialize(store))) 稳定', () {
      final store = parseSchemeDoc('');
      final md = serializeSchemeDoc(store);
      final reparsed = parseSchemeDoc(md);
      expect(reparsed.activeSchemeId, store.activeSchemeId);
      expect(reparsed.schemes, hasLength(store.schemes.length));
      // 题目 ID 序列完全保留。
      final before =
          store.activeScheme.questions.map((q) => q.id).toList();
      final after =
          reparsed.activeScheme.questions.map((q) => q.id).toList();
      expect(after, before);
    });

    test('自定义方案：解析出 activeSchemeId + 自定义题目', () {
      const md = '''# LifeScale 复盘方案

```json
{
  "activeSchemeId": "scheme-custom-light",
  "schemes": [
    {
      "id": "scheme-custom-light",
      "name": "我的轻量复盘",
      "source": "custom",
      "isDefault": false,
      "questions": [
        {"id": "custom-progress", "title": "今天最值得记录的进展？", "required": true, "maxLength": 500, "sortOrder": 1}
      ]
    }
  ]
}
```
''';
      final store = parseSchemeDoc(md);
      expect(store.activeSchemeId, 'scheme-custom-light');
      expect(store.activeScheme.questions.first.id, 'custom-progress');
    });
  });
}
